rust/connlib/tunnel/src/client.rs

```diff
@@ -1017,41 +992,9 @@ impl ClientState {
 
             self.resource_ids.remove(id);
 
-            let Some(gateway_id) = self.resources_gateways.remove(id) else {
+            if self.resources_gateways.remove(id).is_none() {
                 tracing::debug!("No gateway associated with resource");
-                continue;
             };
-
-            let Some(peer) = self.peers.get_mut(&gateway_id) else {
-                continue;
-            };
-
-            // First we remove the id from all allowed ips
-            for (network, resources) in peer
-                .allowed_ips
-                .iter_mut()
-                .filter(|(_, resources)| resources.contains(id))
-            {
-                resources.remove(id);
-
-                if !resources.is_empty() {
-                    continue;
-                }
-
-                // If the allowed_ips doesn't correspond to any resource anymore we
-                // clean up any related translation.
-                peer.translations.remove_by_left(&network.network_address());
-            }
-
-            // We remove all empty allowed ips entry since there's no resource that corresponds to it
-            peer.allowed_ips.retain(|_, r| !r.is_empty());
-
-            // If there's no allowed ip left we remove the whole peer because there's no point on keeping it around
-            if peer.allowed_ips.is_empty() {
-                self.peers.remove(&gateway_id);
-                self.update_site_status_by_gateway(&gateway_id, Status::Unknown);
-                // TODO: should we have a Node::remove_connection?
-            }
```

Comment:
I think this is fine. I think what ends up happening here is that we have a connection to a gateway that we no longer need. We can clean up unused connections at a later point and I think it is probably a feature we should add to snownet and clean them up based on how often we see traffic going through even if we still have a resource mapping.

Reply:
If we keep track of gateways to resources ids like suggested [here](https://github.com/firezone/firezone/issues/5054#issuecomment-2121485144) we can cleanup peers when no resource is around that corresponds to a gateway.

Reply:
Sounds good. I am planning to work on #5054 next.